src/sinks/file.rs

```diff
@@ -0,0 +1,316 @@
+        SinkExt,
+    },
+    template::Template,
+    topology::config::DataType,
+};
+
+use futures::{future, Async, AsyncSink, Sink, StartSend};
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::path::PathBuf;
+use std::time::Instant;
+
+#[derive(Deserialize, Serialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct FileSinkConfig {
+    #[serde(rename = "path")]
+    pub path_template: String,
+    #[serde(rename = "timeout", default = "default_close_timeout_secs")]
+    pub close_timeout_secs: Option<u64>,
+    pub encoding: Option<BasicEncoding>,
+}
+
+fn default_close_timeout_secs() -> Option<u64> {
+    None //means that we don't close files, so no overhead
+}
+
+impl FileSinkConfig {
+    pub fn new(path_template: String) -> Self {
+        Self {
+            path_template,
+            close_timeout_secs: default_close_timeout_secs(),
+            encoding: None,
+        }
+    }
+}
+
+#[typetag::serde(name = "file")]
+impl crate::topology::config::SinkConfig for FileSinkConfig {
+    fn build(&self, acker: Acker) -> Result<(super::RouterSink, super::Healthcheck), String> {
+        let sink = PartitionedFileSink::new(
+            Template::from(&self.path_template[..]),
+            self.close_timeout_secs,
+            self.encoding.clone(),
+        )
+        .stream_ack(acker);
+
+        Ok((Box::new(sink), Box::new(future::ok(()))))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+}
+
+pub struct PartitionedFileSink {
+    path_template: Template,
+    encoding: Option<BasicEncoding>,
+    close_timeout_secs: Option<u64>,
+
+    partitions: HashMap<PathBuf, EmbeddedFileSink>,
+    last_accessed: HashMap<PathBuf, Instant>,
+    closing: Vec<EmbeddedFileSink>,
+}
+
+impl PartitionedFileSink {
+    pub fn new(
+        path_template: Template,
+        close_timeout_secs: Option<u64>,
+        encoding: Option<BasicEncoding>,
+    ) -> Self {
+        PartitionedFileSink {
+            path_template,
+            encoding,
+            close_timeout_secs,
+            partitions: HashMap::new(),
+            last_accessed: HashMap::new(),
+            closing: Vec::new(),
+        }
+    }
+
+    fn collect_old_files(&mut self) {
+        let mut recently_outdated = vec![];
+        for timeout in self.close_timeout_secs {
+            self.last_accessed.retain(|path, time| {
+                if time.elapsed().as_secs() > timeout {
+                    debug!(message = "removing sink", file = ?path);
+                    recently_outdated.push(path.clone());
+                    false
+                } else {
+                    true
+                }
+            });
+        }
+
+        let mut recently_outdated = recently_outdated
+            .into_iter()
+            .map(|ref path| self.partitions.remove(path).unwrap())
+            .collect();
+
+        //it is easier to empty `closing` buffer and put back not closed sinks, see `poll_close`
```

Comment:
I think this is what you are trying to say?

```suggestion
        // it is easier to empty a `closing` buffer and then put it back into the not closed sinks, see `poll_close`
```

Reply:
Almost; sorry for bad grammar.
> and then put it back into the not closed sinks
> and then put back all sinks which are not closed yet